File: tps/entrada e saida/tp03/TP03Q03 - Pilha com Alocação Sequencial em Java/Jogador.java
public class Jogador {

  static final String arquivoCSV = "/tmp/players.csv";
  static final int maxEntradas = 128;
  static class Lista {
    public Jogador j[];

    public Lista() {
      j = new Jogador[0];
    }
    public Lista(int qtd) {
      j = new Jogador[qtd];
    }


    // insere um registro na primeira posicao da Lista 
    public void inserirInicio(Jogador jogador) {
      Jogador novo[] = new Jogador[j.length + 1];
      novo[0] = jogador.clone();
      // remaneja os demais
      for(int i = 1; i < j.length+1; i++) {
        novo[i] = j[i-1].clone();
      }
      j = novo;
    }


    // insere um registro na posicao p da Lista, onde p < n e n o numero de registros
    public void inserir(int posicao, Jogador jogador) {
      Jogador novo[] = new Jogador[j.length + 1];
      // remaneja os registros
      int i = 0;
      while(i < posicao) {
        novo[i] = j[i++].clone();
      }
      novo[posicao] = jogador.clone();
      while(i < j.length) {
        novo[i+1] = j[i++].clone();
      }
      j = novo;
    }


    // insere um registro na ultima posicao da Lista
    public void inserirFim(Jogador jogador) {
      Jogador novo[] = new Jogador[j.length + 1];
      novo[j.length] = jogador.clone();
      for(int i = 0; i < j.length; i++) {
        novo[i] = j[i].clone();
      }
      j = novo;
    }


    // remove e retorna o primeiro registro cadastrado 
    // na Lista e remaneja os demais
    public Jogador removerInicio() {
      Jogador novo[] = new Jogador[j.length - 1];
      Jogador resp = j[0].clone();
      for(int i = 0; i < novo.length; i++) {
        novo[i] = j[i+1].clone();
      }
      j = novo;
      System.out.println("(R) " + resp.getNome());
      return resp;
    }


    //remove e retorna o registro cadastrado na p-esima 
    //posicao da Lista e remaneja os demais. 
    public Jogador remover(int posicao) throws Exception {
      if(posicao >= j.length) throw new Exception("Posicao invalida");
      
      Jogador novo[] = new Jogador[j.length - 1];
      Jogador resp = j[posicao].clone();
      
      int i = 0;
      while(i < posicao) {
        novo[i] = j[i++].clone();
      }
      while(i < novo.length) {
        novo[i] = j[i+1].clone();
        i++;
      }
      j = novo;
      System.out.println("(R) " + resp.getNome());
      return resp;
    }


    // remove e retorna o ultimo registro cadastrado na lista.
    public Jogador removerFim() {
      Jogador novo[] = new Jogador[j.length - 1];
      Jogador resp = j[j.length - 1].clone();
      for(int i = 0; i < j.length-1; i++) {
        novo[i] = j[i].clone();
      }
      j = novo;
      System.out.println("(R) " + resp.getNome());
      return resp;
    }


    public void imprimir() {
      for(int i = 0; i < j.length; i++) {
        System.out.print("[" + i + "] ");
        j[i].imprimir();
      }
    }
  }


  int id;
  String nome;
  int altura;
  int peso;
  String universidade;
  int anoNascimento;
  String cidadeNascimento;
  String estadoNascimento;

  public Jogador() {
  }

  public Jogador(String linha) {
    String campos[] = linha.split(",");
    this.id = Integer.parseInt(campos[0]);
    this.nome = campos[1];
    this.altura = Integer.parseInt(campos[2]);
    this.peso = Integer.parseInt(campos[3]);
    this.universidade = (campos[4].isEmpty()) ? "nao informado" : campos[4];
    this.anoNascimento = Integer.parseInt(campos[5]);
    if (campos.length > 6) {
      this.cidadeNascimento = (campos[6].isEmpty()) ? "nao informado": campos[6];
      if (campos.length < 8) {
        this.estadoNascimento = "nao informado";
      } else {
        this.estadoNascimento = campos[7];
      }
    } else {
      this.cidadeNascimento = "nao informado";
      this.estadoNascimento = "nao informado";
    }
  }

  // id,Player,height,weight,collage,born,birth_city,birth_state
  public int getId() {
    return id;
  }

  public void setId(int id) {
    this.id = id;
  }

  public String getNome() {
    return nome;
  }

  public void setNome(String nome) {
    this.nome = nome;
  }

  public int getAltura() {
    return altura;
  }

  public void setAltura(int altura) {
    this.altura = altura;
  }

  public int getPeso() {
    return peso;
  }

  public void setPeso(int peso) {
    this.peso = peso;
  }

  public void setAnoNascimento(int anoNascimento){
    this.anoNascimento = anoNascimento;
  }

  public int getAnoNascimento(){
    return anoNascimento;
  }

  public String getUniversidade() {
    return universidade;
  }

  public void setUniversidade(String universidade) {
    this.universidade = universidade;
  }

  public String getCidadeNascimento() {
    return cidadeNascimento;
  }

  public void setCidadeNascimento(String cidadeNascimento) {
    this.cidadeNascimento = cidadeNascimento;
  }

  public String getEstadoNascimento() {
    return estadoNascimento;
  }

  public void setEstadoNascimetno(String estadoNascimento) {
    this.estadoNascimento = estadoNascimento;
  }

  public Jogador clone() {
    Jogador novo = new Jogador();
    novo.id = this.id;
    novo.nome = this.nome;
    novo.altura = this.altura;
    novo.peso = this.peso;
    novo.anoNascimento = this.anoNascimento;
    novo.universidade = this.universidade;
    novo.cidadeNascimento = this.cidadeNascimento;
    novo.estadoNascimento = this.estadoNascimento;
    return novo;
  }

  public void imprimir() {
    System.out.println("## " + nome + " ## " + altura + " ## " + peso + " ## " + anoNascimento + " ## "
        + universidade + " ## " + cidadeNascimento + " ## " + estadoNascimento + " ##");
  }

  public String toString() {
    return "[" + id + " ## " + nome + " ## " + altura + " ## " + peso + " ## " + anoNascimento + " ## "
      + universidade + " ## " + cidadeNascimento + " ## " + estadoNascimento + "]";
  }

  
  // retorna String com conteudo da linha correspondente ao 
  // id enviado como parametro
  public static String leLinhaCSV(int id) {
    Arq.openRead(arquivoCSV);
    String resp = "";
    // mover cursor do csv ate a linha do id
    for(int n = 0; n <= id + 1; n++) { 
      resp = Arq.readLine();  // le a linha com id inserido e armazena a string em "linha"
    }
    Arq.close();
    return resp;
  }


  // preenche vetor enviado como parametro com leitura de entrada padrao
  // retorna quantidade de entradas efetuadas
  public static int leituraDeEntrada(int entrada[]) {
    int numEntrada = 0;
    do {
      entrada[numEntrada] = MyIO.readInt();
    } while (entrada[numEntrada++] != -1);
    numEntrada--;   //Desconsiderar linha contendo a palavra FIM
    return numEntrada;
  }


  // preenche vetor enviado como parametro com leitura de entrada padrao
  // retorna quantidade de entradas efetuadas
  public static int leituraDeEntrada(String entrada[]) {
    int numEntrada = 0;
    do {
      entrada[numEntrada] = MyIO.readLine();
    } while (!entrada[numEntrada++].equals("FIM"));
    numEntrada--;   //Desconsiderar linha contendo a palavra FIM
    return numEntrada;
  }




  public static void main(String[] args) throws Exception {
    // array que armazena as linhas de entrada
    int[] entrada = new int[maxEntradas];

    // Primeira leitura da entrada padrao com os ids
    int numEntrada = leituraDeEntrada(entrada);

    // formar array de jogadores    
    Lista lista = new Lista();
    
    for(int i = 0; i < numEntrada; i++) {
      Jogador jogador = new Jogador(leLinhaCSV(entrada[i]));
      lista.inserirFim( jogador );
    }

    // segunda leitura de entrada
    int numOperacoes = MyIO.readInt();
    char operacao;

    for (int i = 0; i < numOperacoes; i++) {
      operacao = MyIO.readChar();
      MyIO.readChar();
      switch(operacao) {
        //case "I*": lista.inserir(MyIO.readInt(), new Jogador( leLinhaCSV(MyIO.readInt()) )); break;
        //case "II": lista.inserirInicio(new Jogador( leLinhaCSV(MyIO.readInt()) )); break;
        case 'I': lista.inserirFim(new Jogador( leLinhaCSV(MyIO.readInt()) )); break;
        //case "R*": lista.remover(MyIO.readInt()); break;
        //case "RI": lista.removerInicio(); break;
        case 'R': lista.removerFim(); break;
        default: System.out.println("operacao invalida"); break;
      }
    }

    lista.imprimir();

  }
}

File: tps/entrada e saida/tp03/TP03Q10 - Pilha com Alocação Flexível em C/jogador.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <err.h>
#include <time.h>

#define TAM_MAX_LINHA 250
#define MAX_ENTRADAS 500
#define arquivoCSV "/tmp/players.csv"

int numComparacoes = 0;
int numMovimentacoes = 0;


typedef struct {
  int id;
  int peso;
  int altura;
  char nome[70];
  char universidade[70];
  int anoNascimento;
  char cidadeNascimento[70];
  char estadoNascimento[70];
} Jogador;


void setJogador(char* strAttr[], Jogador* J) {
  J->id = atoi(strAttr[0]);
  strcpy(J->nome, strAttr[1]);
  J->altura = atoi(strAttr[2]);
  J->peso = atoi(strAttr[3]);
  strcpy(J->universidade, strAttr[4]);
  J->anoNascimento = atoi(strAttr[5]);
  strcpy(J->cidadeNascimento, strAttr[6]);
  if(strAttr[7][0]=='\0') 
    strcpy(J->estadoNascimento, "zzz");
  else strcpy(J->estadoNascimento, strAttr[7]);
}


void imprimir(Jogador *J) {
  char* vazio = "nao informado";
  printf("## %s ## %d ## %d ## %d ## %s ## %s ## %s ##\n",
    //J->id,
    J->nome[0]=='\0' ? vazio : J->nome,
    J->altura,
    J->peso,
    J->anoNascimento,
    J->universidade[0]=='\0' ? vazio : J->universidade,
    J->cidadeNascimento[0]=='\0' ? vazio : J->cidadeNascimento,
    strcmp(J->estadoNascimento,"zzz") == 0 ? vazio : J->estadoNascimento
  );
}


Jogador clone(Jogador *jogador) {
  Jogador retorno;
  retorno.id = jogador->id;
  strcpy(retorno.nome, jogador->nome);
  retorno.altura = jogador->altura;
  retorno.peso = jogador->peso;
  retorno.anoNascimento = jogador->anoNascimento;
  strcpy(retorno.universidade, jogador->universidade);
  strcpy(retorno.cidadeNascimento, jogador->cidadeNascimento);
  strcpy(retorno.estadoNascimento, jogador->estadoNascimento);
  return retorno;
}


// recebe string de uma lihna do csv e coloca os atributos no array
void parseAtributos(char* input, char* output[]) {
  int i = 0;
  strtok(input, "\n"); // retira quebra de linha no final da string
  
  output[i++] = input;

  // loop vai repartindo linha nas ',' e atribuindo a output[]
  while (i < 8 && input && (input = strchr(input, ','))) {
    *input = 0;           
    output[i++] = ++input;
  }
}


// le a lihna do csv correspondente ao id procurado
void lerLinha(FILE* csv, char linha[], int id) {
  rewind(csv);
  for(int n = 0; n <= id + 1; n++) {
    fgets(linha, TAM_MAX_LINHA, csv);
  }
}


// atribui cada atributo do array de strings no objeto Jogador
Jogador ler(FILE *csv, int id) {
  Jogador J;
  char linha[TAM_MAX_LINHA]; // linha a ser lida do csv 
  char* strAttr[8]; // array com os atributos

  lerLinha(csv, linha, id); // le linha do arquivo correspondente ao id
  
  parseAtributos(linha, strAttr); // transforma a string em array de atributos
  
  setJogador(strAttr, &J); // cria o registro jogador a partir do array de atributos
  
  return J; 
}


// faz leitura dos valores a partir da entrada padrao e 
// retorna numero de entrada efetuadas
int leituraDeEntrada(char entrada[][TAM_MAX_LINHA]) {
  int numEntrada = 0; // contador de entradas
  do {
    scanf("%[^\n]%*c", entrada[numEntrada]); // *c ignora ultimo caracter lido
  } while(strcmp(entrada[numEntrada++],"FIM")!=0);
  numEntrada--;  // desconsidera entrada com FIM
  return numEntrada;
}



//TIPO CELULA ===================================================================
typedef struct CelulaDupla {
	Jogador elemento;        // Elemento inserido na celula.
	struct CelulaDupla* prox; // Aponta a celula prox.
   struct CelulaDupla* ant;  // Aponta a celula anterior.
} CelulaDupla;

CelulaDupla* novaCelulaDupla(Jogador* elemento) {
   CelulaDupla* nova = (CelulaDupla*) malloc(sizeof(CelulaDupla));
   nova->elemento = *elemento;
   nova->ant = nova->prox = NULL;
   return nova;
}

//LISTA PROPRIAMENTE DITA =======================================================
CelulaDupla* primeiro;
CelulaDupla* ultimo;


/**
 * Cria uma lista dupla sem elementos (somente no cabeca).
 */
void start () {
  Jogador j;
  primeiro = novaCelulaDupla(&j);
  ultimo = primeiro;
}


/**
 * Insere um elemento na ultima posicao da lista.
 * @param x int elemento a ser inserido.
 */
void inserirFim(Jogador* x) {
   ultimo->prox = novaCelulaDupla(x);
   ultimo->prox->ant = ultimo;
   ultimo = ultimo->prox;
}


/**
 * Remove um elemento da ultima posicao da lista.
 * @return resp int elemento a ser removido.
 */
Jogador removerFim() {
   if (primeiro->elemento.id == ultimo->elemento.id) {
      errx(1, "Erro ao remover (vazia)!");
   } 
   Jogador resp = ultimo->elemento;
   ultimo = ultimo->ant;
   ultimo->prox->ant = NULL;
   free(ultimo->prox);
   ultimo->prox = NULL;
   return resp;
}


/**
 *  Calcula e retorna o tamanho, em numero de elementos, da lista.
 *  @return resp int tamanho
 */
int tamanho() {
   int tamanho = 0; 
   CelulaDupla* i;
   for(i = primeiro; i != ultimo; i = i->prox, tamanho++);
   return tamanho;
}


// Mostra os elementos separados por espacos.
void mostrar() {
  CelulaDupla* i;
  int cont = 0;
  for (i = primeiro->prox; i != NULL; i = i->prox) {
    printf("[%d] ", cont++);
    imprimir(&i->elemento);
  }
}


// MAIN ===============================================================
int main(int argc, char *argv[]) {
  FILE *csv = fopen(arquivoCSV, "r");

  // variavel para armazenar entrada padrao com id's
  char entrada[MAX_ENTRADAS][TAM_MAX_LINHA]; 
  
  // leitura da entrada padrao, retorna quantidade de entradas inseridas
  const int numEntrada = leituraDeEntrada(entrada); 
  
  // iniciando lista
  start();
  
  // criar registros "jogador" a partir das entradas armazenadas
  for (int i = 0; i < numEntrada; i++) {
    Jogador j = ler(csv, atoi(entrada[i])); // atoi(entrada[i]) = id do jogador
    inserirFim(&j);
  }
  
  int numOperacoes, posicao, id; 
  scanf("%d",&numOperacoes);
  
  for(int i = 0; i<numOperacoes; i++) {
    Jogador jogador;
    char operacao[2]; scanf("%s", operacao);

      if (strcmp(operacao,"I") == 0) {
        scanf("%d", &id); 
        jogador = ler(csv, id);
        inserirFim(&jogador);
      } else if (strcmp(operacao,"R") == 0) {
        printf("(R) %s\n", (removerFim()).nome);
      }
      else printf("operacao invalida\n");
  }

  mostrar();

  fclose(csv);
  
  return 0;
}
